src/daft-functions/src/uri/upload.rs

```diff
@@ -28,59 +29,155 @@ impl ScalarUDF for UploadFunction {
                 match data_field.dtype {
-                    DataType::Binary | DataType::FixedSizeBinary(..) | DataType::Utf8 => Ok(Field::new(data_field.name, DataType::Utf8)),
-                    _ => Err(DaftError::TypeError(format!("Expects input to url_upload to be Binary, FixedSizeBinary or String, but received {data_field}"))),
+                    DataType::Binary | DataType::FixedSizeBinary(..) | DataType::Utf8 => (),
+                    _ => return Err(DaftError::TypeError(format!("Expects input to url_upload to be Binary, FixedSizeBinary or String, but received {data_field}"))),
                 }
+                if !location_field.dtype.is_string() {
+                    return Err(DaftError::TypeError(format!(
+                        "Expected location to be string, received: {}",
+                        location_field.dtype
+                    )));
+                }
+                Ok(Field::new(data_field.name, DataType::Utf8))
             }
             _ => Err(DaftError::SchemaMismatch(format!(
-                "Expected 1 input arg, got {}",
+                "Expected 2 input args, got {}",
                 inputs.len()
             ))),
         }
     }
 }
 
+/// Helper function that takes a given folder path, a boolean `is_single_folder` that indicates if we were given a single folder
+/// for uploading files or if we were given a row-specific path, and a set of instantiated folder paths, then creates the folder
+/// if they has not yet been instantiated.
+///
+/// This function also parses the given folder path and trims '/' from the end of the path.
+fn instantiate_and_trim_path(
+    folder_path: &str,
+    is_single_folder: bool,
+    instantiated_folder_paths: &mut HashSet<String>,
+) -> DaftResult<String> {
+    // HACK: Creates folders if running locally. This is a bit of a hack to do it here because we'd rather delegate this to
+    // the appropriate source. However, most sources such as the object stores don't have the concept of "folders".
+    let (source, folder_path) = daft_io::parse_url(folder_path)?;
+    if matches!(source, SourceType::File) {
+        let local_prefixless_folder_path = match folder_path.strip_prefix("file://") {
+            Some(p) => p,
+            None => folder_path.as_ref(),
+        };
+        if is_single_folder {
+            // If we were given a single folder, create a directory at the given path.
+            if instantiated_folder_paths.insert(local_prefixless_folder_path.to_string()) {
+                std::fs::create_dir_all(local_prefixless_folder_path).map_err(|e| {
+                    daft_io::Error::UnableToCreateDir {
+                        path: local_prefixless_folder_path.to_string(),
+                        source: e,
+                    }
+                })?;
+            }
+        } else {
+            // If we were given row-specific paths, then we create directories at the parents of the given paths.
```

Comment:
Why do create directories for the *parents* of the paths in each row for this case?

Reply:
So my thinking was that if a single path was provided to `upload()` we would use this path as a directory, then upload all the files with UUIDs as their names inside this directory:
```
parent
└── directory
    ├── {UUID1}
    ├── ...
    └── {UUIDn}
```

But if each row is given a specific path, then instead of treating the path as a directory prefix, we should upload the file to the path itself. Since the final pathname component is a file and not a directory, only the directories up to the parents should be created.

Does this work as an API or do you have something else in mind?

Reply:
Ahh okay I see, that makes sense to me from a user's perspective. Thanks for clearing it up!